File: CS/WebSite/WebUserControl2.ascx.cs
using System;
using DevExpress.Web.ASPxGridView;
using DevExpress.Web.ASPxEditors;

public partial class WebUserControl2 : System.Web.UI.UserControl {
    protected void Page_Load(object sender, EventArgs e) {

    }

    protected void gvDetail_BeforePerformDataSelect(object sender, EventArgs e){
        Session["MasterID"] = (sender as ASPxGridView).GetMasterRowKeyValue();
    }
    protected void gvDetail_InitNewRow(object sender, DevExpress.Web.Data.ASPxDataInitNewRowEventArgs e){
        Session["MasterID"] = (sender as ASPxGridView).GetMasterRowKeyValue();
        e.NewValues["MasterID"] = Session["MasterID"].ToString();
    }

    protected void GridView_RowInserting(object sender, DevExpress.Web.Data.ASPxDataInsertingEventArgs e){
        ASPxGridView gridView = (ASPxGridView)sender;

        GridDataSource gridDataSource = new GridDataSource();
        gridDataSource.InsertRow(e.NewValues, gridView.SettingsDetail.IsDetailGrid);

        gridView.CancelEdit();
        e.Cancel = true;
    }
    protected void gvDetail_CellEditorInitialize(object sender, ASPxGridViewEditorEventArgs e) {
        ASPxGridView gridView = (ASPxGridView)sender;
        if (e.Column.FieldName == "MasterID") {
            ASPxTextBox textBoxMasterID = (ASPxTextBox)e.Editor;
            if (textBoxMasterID.Value == null)
                textBoxMasterID.Value = gridView.GetMasterRowKeyValue();
        }
    }
}

File: VB/WebSite/WebUserControl2.ascx.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Web.ASPxGridView
Imports DevExpress.Web.ASPxEditors

Partial Public Class WebUserControl2
	Inherits System.Web.UI.UserControl
	Protected Sub Page_Load(ByVal sender As Object, ByVal e As EventArgs)

	End Sub

	Protected Sub gvDetail_BeforePerformDataSelect(ByVal sender As Object, ByVal e As EventArgs)
		Session("MasterID") = (TryCast(sender, ASPxGridView)).GetMasterRowKeyValue()
	End Sub
	Protected Sub gvDetail_InitNewRow(ByVal sender As Object, ByVal e As DevExpress.Web.Data.ASPxDataInitNewRowEventArgs)
                                Session("MasterID") = (TryCast(sender, ASPxGridView)).GetMasterRowKeyValue()
		e.NewValues("MasterID") = Session("MasterID").ToString()
	End Sub

	Protected Sub GridView_RowInserting(ByVal sender As Object, ByVal e As DevExpress.Web.Data.ASPxDataInsertingEventArgs)
		Dim gridView As ASPxGridView = CType(sender, ASPxGridView)

		Dim gridDataSource As New GridDataSource()
		gridDataSource.InsertRow(e.NewValues, gridView.SettingsDetail.IsDetailGrid)

		gridView.CancelEdit()
		e.Cancel = True
	End Sub
	Protected Sub gvDetail_CellEditorInitialize(ByVal sender As Object, ByVal e As ASPxGridViewEditorEventArgs)
		Dim gridView As ASPxGridView = CType(sender, ASPxGridView)
		If e.Column.FieldName = "MasterID" Then
			Dim textBoxMasterID As ASPxTextBox = CType(e.Editor, ASPxTextBox)
			If textBoxMasterID.Value Is Nothing Then
				textBoxMasterID.Value = gridView.GetMasterRowKeyValue()
			End If
		End If
	End Sub
End Class
